### ISSUE TITLE
Cause of errors when running bulk loading jobs?

### ISSUE THREAD

**Issue by btski** [tripal/tripal#490]
Tripal Version: 3
Drupal Version: 7.56
PostgreSQL Version: 9.5.12
PHP Version: 7.0

I'm trying to load sequences using the bulk loader and it's stalling out about halfway through. I would like to cancel/remove jobs from the job menu. Currently, this doesn't appear possible through the job interface itself. Help?

Edit1 - I deleted the associated bulk load template which removed its job page. Is there a way to cancel all/remove all jobs?

Edit2 - Looks like it's just a visual bug. I wasn't able to cancel the running jobs (using: kill PID). Started a new job and it seems to be running fine (prioritizes newest job over queued cancelled jobs). Here's what my job interface currently looks like. I believe 3 and 4 are cancelled. 
![screen shot 2018-06-28 at 12 59 28 pm](https://user-images.githubusercontent.com/32686196/42052023-45e0afce-7ad3-11e8-9c7a-19abdc64d34e.png)

**Comment by btski** [comment]
Just realized that jobs can run in parallel and don't necessarily run chronologically. I've still got jobs listed as running. Unable to tell if visual bug or not. Ran _ps aux_ to check for running jobs by matching PIDs (none found).

**Comment by spficklin** [comment]
Right now Tripal can only update the status of the job the next time it checks for a job.  This only happens if you have it setup to run jobs by cron or if you manually run the job checker on your own.   Can you try and run the ```drush trp-run-jobs```  and see if that updates the status for you?

**Comment by btski** [comment]
Thanks for the suggestion @spficklin. Which module is trp-run-jobs listed in? I don't appear to have it currently installed. My biggest problem now is that bulk load jobs randomly stall/stop and have to be re-launched. The behavior is inconsistent and does not seem linked to user interaction. Presumably, the database is updating (in some way) at regular intervals (that I'm not tracking), which causes problem for running bulk load jobs. Any idea as to what might be causing this issue?

Error report:
```
Notice: Undefined offset: 0 in tripal_bulk_loader_job_describe_args() (line 475 of /var/www/html/sites/all/modules/tripal/tripal_bulk_loader/tripal_bulk_loader.module).
Notice: Trying to get property of non-object in tripal_bulk_loader_job_describe_args() (line 476 of /var/www/html/sites/all/modules/tripal/tripal_bulk_loader/tripal_bulk_loader.module).
Notice: Undefined offset: 0 in tripal_bulk_loader_job_describe_args() (line 476 of /var/www/html/sites/all/modules/tripal/tripal_bulk_loader/tripal_bulk_loader.module).
```
Edit1 - I've tried fragmenting my files with the hope that more will run to completion. I'm currently on my last file of <10,000 sequences and it's stopped anywhere from 3-91%. Most recently was 3%, so I'm trying to immediately restart the job. If bulk load job failure corresponds to some regular schedule, I would expect this most recent attempt to be successful.

**Comment by bradfordcondon** [comment]
`trp-run-jobs`  is described in the tripal module.   https://github.com/tripal/tripal/blob/7.x-3.x/tripal/tripal.drush.inc .  It should be available if Tripal is installed.
  
what error message do you get hwen you run `drush trp-run-jobs`?


edit:

I also see that you are running Drupal 5.10.  Is that correct?  It's [end of life was in 2011](https://www.drupal.org/forum/general/news-and-announcements/2011-01-21/so-long-drupal-5x-end-of-life-announcement), it is no longer supported

**Comment by btski** [comment]
Sorry, it seems I was hasty making the initial post and included the default Drupal version (currently using 7.56). The errors when running `drush trp-run-jobs`:

Directory /home/ubuntu/.drush/cache exists, but is not writable.     [error]
Please check directory permissions.
The drush command 'trp-run-jobs' could not be found.  Run `drush     [error]
cache-clear drush` to clear the commandfile cache if you have
installed new extensions.
A Drupal installation directory could not be found                   [error]

I've modified the /home/ubuntu/.drush/cache directory permissions to a+wr to avoid the first error, but the other two persisted. Reverted directory permissions back to 755. I also attempted to run `drush cache-clear drush`, which ran successfully. Was still unable to get `drush trp-run-jobs` to work.

**Comment by bradfordcondon** [comment]
are you running the command within your drupal site path ie `/var/www/html` for example.

heres what happens if i run that command in a random dir vs the root dir of my drupal website:

```
drush trp-run-jobs
The drush command 'trp-run-jobs' could not be found.  Run `drush cache-clear drush` to clear the commandfile cache if you have  [error]
installed new extensions.
 ✘ chet@Bradfords-MacBook-Pro  ~  cd UTK/tripal
 chet@Bradfords-MacBook-Pro  ~/UTK/tripal  drush trp-run-jobs
2018-07-03 12:33:35
ERROR: Please provide a username (--username argument) for running this job.
Drush command terminated abnormally due to an unrecoverable error.                                                              [error]
 ✘ chet@Bradfords-MacBook-Pro  ~/UTK/tripal 


```


drush basically does cd ../ until it finds your root directly and looks at your sites/defualt/settings.php to know your site's info such as the db password etc.  So if you run the command outside of the site it will fail.

**Comment by btski** [comment]
That would do it. Didn't cross my mind to switch to the Drupal root directory. After running `drush trp-run-jobs`:

2018-07-03 17:12:37: There are 0 jobs queued.
And the "stalled" jobs have been changed to erros.

![screen shot 2018-07-03 at 12 13 58 pm](https://user-images.githubusercontent.com/32686196/42234616-be6e1202-7eba-11e8-8c29-3243842544cc.png)

**Comment by bradfordcondon** [comment]
ok great.  So can i mark this issue closed or do you have further questions?

**Comment by btski** [comment]
I was less concerned about the visual bug (corrected by running `drush trp-run-jobs`) and more concerned with what was causing all the errors. Any ideas? It's not a big problem, just an annoyance when I have to run the same job half a dozen times over before it completes entirely.

Edit1 - My perception of the problem changed when I realized that the jobs weren't running (as had been shown on the jobs page).

**Comment by bradfordcondon** [comment]
oh, i see.

```

Notice: Undefined offset: 0 in tripal_bulk_loader_job_describe_args() (line 475 of /var/www/html/sites/all/modules/tripal/tripal_bulk_loader/tripal_bulk_loader.module).
Notice: Trying to get property of non-object in tripal_bulk_loader_job_describe_args() (line 476 of /var/www/html/sites/all/modules/tripal/tripal_bulk_loader/tripal_bulk_loader.module).
Notice: Undefined offset: 0 in tripal_bulk_loader_job_describe_args() (line 476 of /var/www/html/sites/all/modules/tripal/tripal_bulk_loader/tripal_bulk_loader.module).
```


The error you supplied is here:

https://github.com/tripal/tripal/blob/b89cafa81fb50daa95ffd308c4a51a8ecde570bb/tripal_bulk_loader/tripal_bulk_loader.module#L470-L480

Looks like `$args[0]` doesnt exist, which is described as: 
>The arguments passed into tripal_add_job()

So... it looks to me like your bulk loading job is somehow being launched without any arguments.




this is your only error for the failed jobs?  Can you rerun a bulk loading job but submit the job with drush trp-run-jobs?  This lets you see the console and likely a better error message.

Also can you export your bulk loader and share with a sample upload file?  That will also help to troubleshoot.

**Comment by btski** [comment]
That's the only error that I've seen via the GUI. But I'm under the impression that the error message is an artifact (unrelated to the issue, just a problem encountered when loading a summary page). I'll rerun a few jobs through drush trp-run-jobs and report back.

Edit1 - I just ran three files, each between 8,000 and 10,000 sequences long. In my previous attempts, each had to be loaded multiple times due to errors. However, when ran using `drush trp-run-jobs`, all three completed successfully.

Conclusion - I would caution against running jobs through the console.

**Comment by bradfordcondon** [comment]
> But I'm under the impression that the error message is an artifact 

I'd agree. the true error will be in the console when manually running the job via `drush trp-run-jobs`

**Comment by btski** [comment]
Thanks for the help @bradfordcondon and @spficklin. There seems to be a bug in the jobs GUI, but it can be circumvented easily through command line.

**Comment by bradfordcondon** [comment]
meaning your jobs ran successfully but the GUI reported them as failed?  If so thats a bug we need to fix.

Can you provide a little more info?

**Comment by btski** [comment]
When using `drush trp-run-jobs`, the GUI correctly matched the command line output (jobs ran to completion). However, when using just GUI "Execute", the jobs would often error (randomly) and be listed as "Running" despite being active. The processes seem to be distinct. While `drush trp-run-jobs` runs linearly in chronological task submission order, the Execute command updates all three simultaneously.

Launching three bulk load files...
via "Execute" the progress might look something like:

<table class="tg">
  <tr>
    <td></td>
    <td>Time point 1</td>
    <td>Time point 2</td>
    <td>Time point 3</td>
    <td>Time point 4</td>
  </tr>
  <tr>
    <th>Job1</th>
    <td>25%</td>
    <td>50%</td>
    <td>60% (broke)</td>
    <td>60%</td>
  </tr>
  <tr>
    <th>Job2</th>
    <td>15% (broke)</td>
    <td>15%</td>
    <td>15%</td>
    <td>15%</td>
  </tr>
  <tr>
    <th>Job3</th>
    <td>25%</td>
    <td>50%</td>
    <td>75%</td>
    <td>100%</td>
  </tr>
</table>

Note - when the running job broke it was still listed as "Running" in the GUI. I believe displayed progress is accurate, but the jobs were not reported in error and the errors were frequent.

via `drush trp-run-jobs` the progress would look like:

<table class="tg">
  <tr>
    <td></td>
    <td>Time point 1</td>
    <td>Time point 2</td>
    <td>Time point 3</td>
    <td>Time point 4</td>
  </tr>
  <tr>
    <th>Job1</th>
    <td>75%</td>
    <td>100%</td>
    <td>100%</td>
    <td>100%</td>
  </tr>
  <tr>
    <th>Job2</th>
    <td>0%</td>
    <td>50%</td>
    <td>100%</td>
    <td>100%</td>
  </tr>
  <tr>
    <th>Job3</th>
    <td>0%</td>
    <td>0%</td>
    <td>25%</td>
    <td>100%</td>
  </tr>
</table>

Note - jobs were completely loaded and listed as "Completed" in the GUI.

**Comment by bradfordcondon** [comment]
thanks so much for the detailed response.

Is it possible that the jobs are dependent on the previously submitted job?  IE job 2 wont work until job 1 finishes?  That would explain this behavior...

**Comment by btski** [comment]
They are not dependent. Even after fully loading the jobs, a rerun with "Execute" can result in failure (will test again to confirm). This is not the case with `drush trp-run-jobs`.

Edit1 - Closest example from memory: Had one of my jobs run to 91%. Re-ran and it ended at 3%. Re-ran again and it made it to 52%. If these errors were due to dependencies (ordering), you should always see positive progress. Plus I stopped running them simultaneously when I started getting errors. This may be a more accurate representation of my experience (and why I stopped running multiple jobs at once).

<table class="tg">
  <tr>
    <td></td>
    <td>Time point 1</td>
    <td>Time point 2</td>
    <td>Time point 3</td>
    <td>Time point 4</td>
  </tr>
  <tr>
    <th>Job1</th>
    <td>14% (break)</td>
    <td>14%</td>
    <td>14%</td>
    <td>14%</td>
  </tr>
  <tr>
    <th>Job2</th>
    <td>13% (broke)</td>
    <td>13%</td>
    <td>13%</td>
    <td>13%</td>
  </tr>
  <tr>
    <th>Job3</th>
    <td>13% (broke)</td>
    <td>13%</td>
    <td>13%</td>
    <td>13%</td>
  </tr>
</table>

I can't entirely remember if this is accurate. Unfortunately, the progress was overwritten when I re-ran the jobs. I think I recall all jobs breaking at about the same time with similar progression.

Edit2 - Most recent re-run with "Execute" completed successfully according to GUI. The job had previously been completed successfully.

**Comment by spficklin** [comment]
@btski yes, you do not want to run long-running jobs via the GUI.  The reason we have the Job's system is because the GUI times out, so long-running jobs should really be run via the console, or via a cron setup (See this document: http://tripal.info/tutorials/v3.x/installation/job-automation).  

The ability to run a job via the GUI was added later and should only be used when you know a job can complete in a reasonable time frame.

**Comment by spficklin** [comment]
But having said that.... we should still fix the problem....

**Comment by laceysanderson** [comment]
I'm closing this as there has been no movement in over a year. Furthermore, the conversation has become long enough that I'm no longer certain what needs to be fixed. If this is still a problem, please create a new issue describing the problem and we'll take another look at it.
